parser/src/sema/semantic_analysis.rs

```diff
@@ -0,0 +1,1607 @@
+                self.locals.insert(
+                    namespaced_name,
+                    BindingType::TraceParam(TraceRef {
+                        span: trace_binding.span,
+                        segment: trace_segment.id,
+                        index: trace_binding.offset,
+                        size: trace_binding.size,
+                    }),
+                );
+            }
+        }
+
+        // Visit all of the statements in the body
+        self.visit_mut_statement_block(&mut function.body)?;
+
+        // Update the dependency graph for this function
+        let current_item = QualifiedIdentifier::new(
+            self.current_module.unwrap(),
+            NamespacedIdentifier::Function(function.name),
+        );
+        for (referenced_item, ref_type) in self.referenced.iter() {
+            let referenced_item = self.deps.add_node(*referenced_item);
+            self.deps.add_edge(current_item, referenced_item, *ref_type);
+        }
+
+        // Restore the original references metadata
+        self.referenced = referenced;
+        // Restore the original lexical scope
+        self.locals = prev;
+        // Disallow constraints
+        self.allowed_constraints = AllowedConstraintsType::None;
+
+        ControlFlow::Continue(())
+    }
+
+    fn visit_mut_boundary_constraints(
+        &mut self,
+        body: &mut Vec<Statement>,
+    ) -> ControlFlow<ModuleError> {
+        // Only allow boundary constraints in this context
+        self.allowed_constraints = AllowedConstraintsType::Boundary;
+        // Save the current bindings set, as we're entering a new lexical scope
+        let prev = self.locals.clone();
+        // Visit all of the statements, check variable usage, and track referenced imports
+        self.visit_mut_statement_block(body)?;
+        // Restore the original lexical scope
+        self.locals = prev;
+        // Disallow any constraints
+        self.allowed_constraints = AllowedConstraintsType::None;
+
+        ControlFlow::Continue(())
+    }
+
+    fn visit_mut_integrity_constraints(
+        &mut self,
+        body: &mut Vec<Statement>,
+    ) -> ControlFlow<ModuleError> {
+        // Only allow integrity constraints in this context
+        self.allowed_constraints = AllowedConstraintsType::Integrity;
+        // Save the current bindings set, as we're entering a new lexical scope
+        let prev = self.locals.clone();
+        // Visit all of the statements, check variable usage, and track referenced imports
+        self.visit_mut_statement_block(body)?;
+        // Restore the original lexical scope
+        self.locals = prev;
+        // Disallow any constraints
+        self.allowed_constraints = AllowedConstraintsType::None;
+
+        ControlFlow::Continue(())
+    }
+
+    // Visit scalar constraints and ensure that they are valid semantically, and have correct types
+    fn visit_mut_enforce(&mut self, expr: &mut ScalarExpr) -> ControlFlow<ModuleError> {
```

Comment:
Yeah, it's been on my TODO list to clean up this module a bit. I do prefer to keep the `SemanticAnalysis` implementation in a single file, as it makes navigating it much easier, but some of the auxilary types and functions can be moved out, and even within the implementation some of the functions definitely need to be broken up/factored out. I've already done some cleanup in my PR for the IR lowering, so it might make more sense to handle the clean up there since they need to be merged at the same time anyway, but if you think it would make review easier I can do some of the clean up here first and rebase on top of it.

Reply:
Hmm no, let's leave it as it is if you've already done some of that. I'll just defer a more thorough review of the semantic analysis to the next PR